### ISSUE TITLE
Omniauth Devise Token Auth with Rails - API mode  Integrate with ember

### ISSUE THREAD

**Issue by halleyrv** [omniauth/omniauth#911]
Hi I integrate devise_auth_token with omniauth, but when  for it to work I had to enable the midleware Session and Cokie. While it manages to authenticate and save in my users table the problem is that I do not have to recover the email and the token since as I said so that the validation can work I had to add midleware

The problem is that first omniauth does not respond json, what it does is show me a facebook login window and once it authenticates it saves my user if it does not exist and then redirects me to the next path

`http://localhost:3000/api/v1/auth/facebook/callback#_=_`

Here I leave the code If someone could help me integrate it with ember js in front end? How I can recovery token or data for user authenticated.

This is my code

first i add gem in gemfile

```
gem 'devise_token_auth'
gem 'omniauth-facebook'
```
then I create initializer in config/initializers/omniauth.rb and puts this code

```
Rails.application.config.middleware.use OmniAuth::Builder do
  provider :facebook,      ENV['FACEBOOK_KEY'], ENV['FACEBOOK_SECRET']
end 
```

In my route i put this code

```
namespace :api do
  	namespace :v1 do
  		mount_devise_token_auth_for 'User', at: 'auth'
  	end	
 end
```

In config/application.rb I add this midleware

```
config.middleware.insert_after ActiveRecord::Migration::CheckPending, ActionDispatch::Cookies
config.middleware.insert_after ActionDispatch::Cookies, ActionDispatch::Session::CookieStore
```

In my model User i puts this code

```
class User < ActiveRecord::Base
  # Include default devise modules.
  devise :database_authenticatable, :registerable,
          :recoverable, :rememberable, :trackable, :validatable,
          :omniauthable, :omniauth_providers => [:facebook]
          #,
          #:confirmable
  include DeviseTokenAuth::Concerns::User
end
```
In my application controller

```
class ApplicationController < ActionController::API
  before_action :authenticate_user!, unless: :devise_controller?
  include DeviseTokenAuth::Concerns::SetUserByToken

end
```

**Comment by mltsy** [comment]
I'm not sure I understand what you're trying to do, but...

OmniAuth isn't really designed for client-side apps.  What you would have to do is redirect the user to the omniauth path `/auth/:provider` as you have done, and in the callback, load the client-side app, with whatever auth properties you want to give it.  But you shouldn't rely on any client-editable data (unencrypted cookies, or in-memory javascript variables, etc) for proof of authentication.

What I have done in the past is make the user sign into the API server with their browser, and have the callback URL set session variables which are stored in an encrypted session cookie in the browser, then redirect them to (or serve directly) the client-side app without any auth details, but have the client-side app make it's own requests to the API, which automatically include the user's session cookie (because the user has a current session with the API server).  In my case, the first request it makes is to request the "current user" to ensure the client is logged in and get the name and user details to display :smile:

**Comment by siamx** [comment]
I have to same issue currently.
So, is there future plans to integrate omniauth with rails api?
How did you sir managed to make it work?
Is there any neat/nifty workarounds to make omniauth with rails api and separated front-end server?

**Comment by alexventuraio** [comment]
I also would like to know the best approach to login users using GitHub account from a ReactJS and a Rails API applications. Are there any suggested source where to start religion or looking at?
Thanks in advance!

**Comment by errawn** [comment]
> I also would like to know the best approach to login users using GitHub account from a ReactJS and a Rails API applications. Are there any suggested source where to start reading or looking at?
> Thanks in advance!

have you figured this out? I'm trying to integrate it also to SPA

**Comment by alexventuraio** [comment]
> > I also would like to know the best approach to login users using GitHub account from a ReactJS and a Rails API applications. Are there any suggested source where to start reading or looking at?
> > Thanks in advance!
> 
> have you figured this out? I'm trying to integrate it also to SPA

No, I haven't found a way to do so, for now, I'm using another omniauth provider instead of Github.

**Comment by errawn** [comment]
> instead

HI! which provider?

**Comment by errawn** [comment]
> > > I also would like to know the best approach to login users using GitHub account from a ReactJS and a Rails API applications. Are there any suggested source where to start reading or looking at?
> > > Thanks in advance!
> > 
> > 
> > have you figured this out? I'm trying to integrate it also to SPA
> 
> No, I haven't found a way to do so, for now, I'm using another omniauth provider instead of Github.

Hi! Which provider?
